File: CSC310-Cpp/_solutions/Module14-String-Algorithms/debug01/naive_match.cpp
// naive_match.cpp
// Naive string matching (CLRS 32.1): tries every shift s of the pattern
// against the text and reports each shift where all m characters agree.
#include <iostream>
#include <string>
#include <vector>

std::vector<int> naiveMatch(const std::string& text, const std::string& pattern) {
    const int n = static_cast<int>(text.size());
    const int m = static_cast<int>(pattern.size());
    std::vector<int> shifts;
    for (int s = 0; s <= n - m; ++s) {
        int k = 0;
        while (k < m && text.at(s + k) == pattern.at(k)) {
            ++k;
        }
        if (k == m) {
            shifts.push_back(s);
        }
    }
    return shifts;
}

void report(const std::string& text, const std::string& pattern) {
    std::vector<int> shifts = naiveMatch(text, pattern);
    std::cout << "\"" << pattern << "\" in \"" << text << "\": ";
    if (shifts.empty()) {
        std::cout << "no match\n";
        return;
    }
    std::cout << shifts.size() << " match" << (shifts.size() == 1 ? "" : "es") << " at shift"
              << (shifts.size() == 1 ? "" : "s");
    for (int s : shifts) {
        std::cout << ' ' << s;
    }
    std::cout << '\n';
}

int main() {
    report("abracadabra", "abra");
    report("aaaaa", "aa");
    report("hello world", "world");
    report("mississippi", "issi");
    report("abc", "abcd");
    report("banana", "nan");
    return 0;
}

File: CSC310-Python/Module14-String-Algorithms/debug01/naive_match.py
"""Naive string matching (CLRS Section 32.1).

naive_match(text, pattern) implements NAIVE-STRING-MATCHER: for every
shift s at which the pattern still fits inside the text, compare the
pattern with the text characters starting at s and record the shift if
every character agrees.  The main block runs seven text/pattern pairs and
prints the shifts found.
"""


def naive_match(text, pattern):
    n = len(text)
    m = len(pattern)
    shifts = []
    for s in range(n - m):
        j = 0
        while j < m and text[s + j] == pattern[j]:
            j += 1
        if j == m:
            shifts.append(s)
    return shifts


def describe(text, pattern):
    shifts = naive_match(text, pattern)
    if not shifts:
        return f'"{pattern}" in "{text}": no match'
    label = "match" if len(shifts) == 1 else "matches"
    where = " ".join(str(s) for s in shifts)
    plural = "shift" if len(shifts) == 1 else "shifts"
    return f'"{pattern}" in "{text}": {len(shifts)} {label} at {plural} {where}'


def main():
    cases = [
        ("abracadabra", "abra"),
        ("aaaaa", "aa"),
        ("hello world", "world"),
        ("mississippi", "issi"),
        ("abc", "abcd"),
        ("banana", "nan"),
        ("abc", "abc"),
    ]
    for text, pattern in cases:
        print(describe(text, pattern))


if __name__ == "__main__":
    main()
